raster-foundry/raster-foundry
# Scopes integration testing

## Overview

This PR adds an integration test to verify that we've attached scope validation to routes in accordance with what we planned to do. 

TODO:

- [x] get a bearer token from the API
- [x] get the testing table from the csv instead of from hard coding two rows of it
- [x] make sbtExplicitDependencies happy
- ~wire up to run in ci from `docker-compose.test.yml`~ (deferred -- azavea/raster-foundry-platform#948)

### Checklist


### Notes

~WIP PR up at this stage even though there's so little here for a specific reason -- I think I've done something pretty clever, and I can't tell if I love it because it's clever or because it's actually useful. One problem I was expecting to run into was: for requests that _are_ allowed, I was going to be messing with a ton of database state, and subsequent runs might have been a problem (e.g. should i keep picking new projects to delete? that means i need to get the list of projects every time. that's not great!). I was also concerned about what would happen if I didn't want to bother with that -- for example, `DELETE` to a project that doesn't exist will return a 403 from authorization failing on the object itself, which makes status code insufficient for determining whether the integration test was successful. I was aiming for a way to evaluate whether authorization would succeed without having to execute the rest of the routing logic, and I think that's what I've done in fdd121c. If people hate it I'll get rid of it but I think it's going to save me a pretty huge headache tomorrow (and should also make the integration tests much faster, since requests will skip even deserializing the body and immediately report their scoped authorization results instead).~

## Testing Instructions

- fire up sbt with some environment variables set: `cd app-backend`, then `SCOPE_IT_API_HOST='http://localhost:9100/api' SCOPE_IT_REFRESH_TOKEN=<a dev refresh token> SCOPE_IT_CSV_LOCATION=$(pwd)/api-it/src/test/resources/endpoint-scopes.csv sbt`
- assemble api server
- somewhere else -- `docker-compose up -d nginx-api api-server`
- wait for the api server to be ready
- from sbt: `apiIntegrationTest/test`
- it will fail -- but confirm that where it reports that it failed is a few records into the csv (the csv is in `api-it/src/test/resources/`)

Closes raster-foundry/annotate#532
